core/src/Telemetry/TelemetryManager.cs

```diff
@@ -0,0 +1,243 @@
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+// 
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Identity.Core.Telemetry
+{
+    internal class TelemetryManager : ITelemetryManager, ITelemetry
+    {
+        private readonly object _lockObj = new object();
+        private ITelemetryReceiver _telemetryReceiver;
+
+        public TelemetryManager(ITelemetryReceiver telemetryReceiver = null)
+        {
+            _telemetryReceiver = telemetryReceiver;
+        }
+
+        public ITelemetryReceiver TelemetryReceiver
+        {
+            get
+            {
+                lock (_lockObj)
+                {
+                    return _telemetryReceiver;
+                }
+            }
+            set
+            {
+                lock (_lockObj)
+                {
+                    _telemetryReceiver = value;
+                }
+            }
+        }
+
+        private bool HasReceiver()
+        {
+            lock (_lockObj)
+            {
+                return _telemetryReceiver != null;
+            }
+        }
+
+        /// <inheritdoc />
+        public string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        public TelemetryHelper CreateTelemetryHelper(
+            string requestId,
+            string clientId,
+            EventBase eventToStart,
+            EventBase eventToEnd = null,
+            bool shouldFlush = false)
+        {
+            return new TelemetryHelper(
+                this,
+                requestId,
+                clientId,
+                eventToStart,
+                eventToEnd ?? eventToStart,
+                shouldFlush);
+        }
+
+        internal readonly ConcurrentDictionary<Tuple<string, string>, EventBase> EventsInProgress = new ConcurrentDictionary<Tuple<string, string>, EventBase>();
+
+        internal readonly ConcurrentDictionary<string, List<EventBase>> CompletedEvents = new ConcurrentDictionary<string, List<EventBase>>();
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            if (!HasReceiver() || string.IsNullOrWhiteSpace(requestId))
+            {
+                return;
+            }
+
+            EventsInProgress[new Tuple<string, string>(requestId, eventToStart[EventBase.EventNameKey])] = eventToStart;
```

Comment:
Would it be a lot of work to replace this tuple with a class? These are so difficult to read...

Reply:
I agree.  Let me look into what we can do here...

Reply:
Fixed